ortb/clone_test.go

```diff
@@ -536,6 +536,326 @@ func TestCloneUser(t *testing.T) {
+		}
+		result := CloneSource(given)
+		assert.Equal(t, given, result, "equality")
+		assert.NotSame(t, given, result, "pointer")
+		assert.NotSame(t, given.SChain, result.SChain, "schain")
+		assert.NotSame(t, given.Ext, result.Ext, "ext")
+	})
+
+	t.Run("assumptions", func(t *testing.T) {
+		assert.ElementsMatch(t, discoverPointerFields(reflect.TypeOf(openrtb2.Source{})),
+			[]string{
+				"SChain",
+				"Ext",
+			})
+	})
+}
+
+func TestCloneSChain(t *testing.T) {
+	t.Run("nil", func(t *testing.T) {
+		result := CloneSource(nil)
+		assert.Nil(t, result)
+	})
+
+	t.Run("empty", func(t *testing.T) {
+		given := &openrtb2.SupplyChain{}
+		result := CloneSChain(given)
+		assert.Empty(t, result)
+		assert.NotSame(t, given, result)
+	})
+
+	t.Run("populated", func(t *testing.T) {
+		given := &openrtb2.SupplyChain{
+			Complete: 1, Nodes: []openrtb2.SupplyChainNode{{ASI: "asi", Ext: json.RawMessage(`{"anyField":1}`)}}}
+		result := CloneSChain(given)
+		assert.Equal(t, given, result, "equality")
+		assert.NotSame(t, given, result, "pointer")
+		assert.NotSame(t, given.Nodes, result.Nodes, "nodes")
+		assert.NotSame(t, given.Ext, result.Ext, "ext")
+	})
+
+	t.Run("assumptions", func(t *testing.T) {
+		assert.ElementsMatch(t, discoverPointerFields(reflect.TypeOf(openrtb2.SupplyChain{})),
+			[]string{
+				"Nodes",
+				"Ext",
+			})
+	})
+}
+
+func TestCloseSupplyChainNodes(t *testing.T) {
+	var np *int8
+	var n int8
+	n = 1
+	np = &n
+	t.Run("nil", func(t *testing.T) {
+		result := CloneSupplyChainNodes(nil)
+		assert.Nil(t, result)
+	})
+
+	t.Run("empty", func(t *testing.T) {
+		given := []openrtb2.SupplyChainNode{}
+		result := CloneSupplyChainNodes(given)
+		assert.Empty(t, result)
+		assert.NotSame(t, given, result)
+	})
+
+	t.Run("one", func(t *testing.T) {
+		given := []openrtb2.SupplyChainNode{
+			{ASI: "asi", HP: np, Ext: json.RawMessage(`{"anyField":1}`)},
+		}
+		result := CloneSupplyChainNodes(given)
+		assert.Equal(t, given, result, "equality")
+		assert.NotSame(t, given[0], result[0], "item-pointer")
+		assert.NotSame(t, given[0].HP, result[0].HP, "item-pointer-hp")
+		assert.NotSame(t, given[0].Ext, result[0].Ext, "item-pointer-ext")
+	})
+
+	t.Run("many", func(t *testing.T) {
+		given := []openrtb2.SupplyChainNode{
+			{ASI: "asi", HP: np, Ext: json.RawMessage(`{"anyField":1}`)},
+			{ASI: "asi", HP: np, Ext: json.RawMessage(`{"anyField":1}`)},
+			{ASI: "asi", HP: np, Ext: json.RawMessage(`{"anyField":1}`)},
+		}
+		result := CloneSupplyChainNodes(given)
+		assert.Equal(t, given, result, "equality")
+		assert.NotSame(t, given[0], result[0], "item0-pointer")
+		assert.NotSame(t, given[0].Ext, result[0].Ext, "item0-pointer-ext")
+		assert.NotSame(t, given[0].HP, result[0].HP, "item0-pointer-hp")
+		assert.NotSame(t, given[1], result[1], "item1-pointer")
+		assert.NotSame(t, given[1].Ext, result[1].Ext, "item1-pointer-ext")
+		assert.NotSame(t, given[1].HP, result[1].HP, "item1-pointer-hp")
+	})
```

Comment:
I think we're missing a couple of asserts here:
```
assert.NotSame(t, given[2].Ext, result[2].Ext, "item2-pointer-ext")
assert.NotSame(t, given[2].HP, result[2].HP, "item2-pointer-hp")
```

Reply:
Yes, added